### PR TITLE
[RSDK-556] remove width/height from GetImage response and create cust…

### PR DESCRIPTION
**Pull request by gvaradarajan**
…om header for raw RGBA data

NOTE: NOT TO BE MERGED UNTIL CORRESPONDING API CHANGE

### CODE REVIEW

**Review by tessavitabile — COMMENTED** [review]
Nice job adding the custom header! 

I'm a little worried this PR is doing too many things at once. I think the main purpose of this PR is to remove width/height from the GetImage response, which requires creating the custom header for raw RGBA (and also allows us to change the internal representation to NRGBA?). I don't understand how changing how lazy is requested relates to the other changes, and it's muddling my understanding of the PR.

**Review by edaniels — CHANGES_REQUESTED** [review]
the rgba stuff looks fine but I don't understand what has been done with lazy. would prefer we just do the rgba stuff here for now and do any improvements/changes to lazy elsewhere for the sake of speed.

File: rimage/depth_map.go [hunk 1]
```diff
@@ -117,28 +116,21 @@ func (dm *DepthMap) SubImage(r image.Rectangle) *DepthMap {
...
-		return ConvertImageToDepthMap(decoded)
 	case *DepthMap:
 		return ii, nil
 	case *imageWithDepth:
 		return ii.Depth, nil
 	case *image.Gray16:
 		return gray16ToDepthMap(ii), nil
-	case *image.NRGBA64:
-		return nrgba64ToDepthMap(ii), nil
+	case *image.NRGBA:
```
[the remark this answers is not in the document]
Comment (bhaney) [reply rimage/depth_map.go:125]: From what I understand, we will be going back to using a raw Depth mime type as well. We were using NRGBA64 for both raw Depth and raw Color (since depth needs 16bit pixels, we changed the color channels to also use 16 bit). 
Switching back to NRGBA means color is 8-bit again (which is more standard, and it will be easier for users to understand the custom rgba mime type)

In another PR, we will being back the `.dat` files and the custom Depth mime type. But that isn't necessary in this PR, as we pretty much just use PNG for most of our depth images.

File: components/camera/server.go [hunk 2]
```diff
@@ -53,8 +53,9 @@ func (s *subtypeServer) GetImage(
 		return nil, err
 	}
 
-	if req.MimeType == "" {
-		req.MimeType = utils.MimeTypeRawRGBALazy
+	err = utils.CheckSupportedMimeType(req.MimeType)
```
[the remark this answers is not in the document]
Comment (gvaradarajan) [reply components/camera/server.go:56]: I struggled to find a suitable default because I didn't want to default to something that used our custom header, but bringing back the lazy mime type should address that

File: rimage/depth_map.go (hunk 1 above)
Comment (bhaney) [inline rimage/depth_map.go:125]: Actually- I think we should completely remove this case. Since NRGBA encode colors in 8-bit, a 8-bit color image will not convert properly into a depth map. with only 2^8 = 256 numbers available, the maximum depth the color image could encode is 256 mm. So if a depth image was converted to a color NRGBA, probably some destructive conversion took place.

So I would say let's remove this case. We will eventually replace it in another PR with the raw depth type

File: rimage/depth_map.go [hunk 3]
```diff
@@ -117,28 +116,21 @@ func (dm *DepthMap) SubImage(r image.Rectangle) *DepthMap {
...
+		return nrgbaToDepthMap(ii), nil
 	default:
 		return nil, errors.Errorf("don't know how to make DepthMap from %T", img)
 	}
 }
 
-// nrgba64ToDepthMap creates a Depthmap from an image.NRGBA64.
-func nrgba64ToDepthMap(img *image.NRGBA64) *DepthMap {
+// nrgba64ToDepthMap creates a Depthmap from an image.NRGBA.
+func nrgbaToDepthMap(img *image.NRGBA) *DepthMap {
```
Comment (bhaney) [inline rimage/depth_map.go:133]: I think this function should be completely removed

File: rimage/image_file.go [hunk 4]
```diff
@@ -209,19 +206,34 @@ func EncodeImage(ctx context.Context, img image.Image, mimeType string) ([]byte,
...
+		binary.BigEndian.PutUint32(widthBytes, uint32(bounds.Dx()))
+		binary.BigEndian.PutUint32(heightBytes, uint32(bounds.Dy()))
+		buf.Write(widthBytes)
+		buf.Write(heightBytes)
+		imgStruct, ok := img.(*image.NRGBA)
+		if !ok {
+			_, ok := img.(*Image)
+			if !ok {
+				return nil, errors.Errorf("underlying struct for image not NRGBA or rimage.Image, detected %T", img)
+			}
```
Comment (bhaney) [inline rimage/image_file.go:232]: You don't need to error here. draw.Draw is smart. It can take any image that fulfills the image.Image interface, and write it properly into `imgStruct`. So rather than erroring, just let draw.Draw do its thing.

Comment (bhaney) [reply rimage/image_file.go:232]: If you wanted to, you don't even need to do a check for underlying type. draw.Draw can take care of it.

**Review by bhaney — CHANGES_REQUESTED** [review]
The encoding is easier than you think -- draw.Draw can convert any image.Image into the right NRGBA image using the struct's color model to do the conversion under the hood. 

I also agree with other commenters that a lot of the lazy encoding stuff should be left in, especially checking the mime type of lazy images in the encoding.

File: rimage/depth_map.go (hunk 3 above)
Comment (bhaney) [reply rimage/depth_map.go:133]: Since NRGBA encode colors in 8-bit, a 8-bit color image will not convert properly into a depth map. with only 2^8 = 256 numbers available, the maximum depth the color image could encode is 256 mm. So if a depth image was converted to a color NRGBA, probably some destructive conversion took place.

So I would say let's remove this case. We will eventually replace it in another PR with the raw depth type

File: components/camera/client.go [hunk 5]
```diff
@@ -50,7 +50,7 @@ func NewClientFromConn(ctx context.Context, conn rpc.ClientConn, name string, lo
 func (c *client) Read(ctx context.Context) (image.Image, func(), error) {
 	ctx, span := trace.StartSpan(ctx, "camera::client::Read")
 	defer span.End()
-	mimeType := gostream.MIMETypeHint(ctx, utils.MimeTypeRawRGBALazy)
+	mimeType := gostream.MIMETypeHint(ctx, utils.MimeTypeRawRGBA)
```
Comment (edaniels) [inline components/camera/client.go:53]: shouldn't this still be lazy?

File: components/camera/server_test.go [hunk 6]
```diff
@@ -84,22 +84,19 @@ func TestServer(t *testing.T) {
 	injectCamera.ProjectorFunc = func(ctx context.Context) (transform.Projector, error) {
 		return projA, nil
 	}
-	wooMIME := "image/woohoo"
 	injectCamera.StreamFunc = func(ctx context.Context, errHandlers ...gostream.ErrorHandler) (gostream.VideoStream, error) {
 		return gostream.NewEmbeddedVideoStreamFromReader(gostream.VideoReaderFunc(func(ctx context.Context) (image.Image, func(), error) {
 			imageReleasedMu.Lock()
 			imageReleased = true
 			imageReleasedMu.Unlock()
-			mimeType, _ := utils.CheckLazyMIMEType(gostream.MIMETypeHint(ctx, utils.MimeTypeRawRGBA))
```
Comment (edaniels) [inline components/camera/server_test.go:93]: why was this removed?

File: rimage/lazy_encoded.go [hunk 7]
```diff
@@ -24,29 +22,36 @@ type LazyEncodedImage struct {
...
 
+type decodeSetting byte
+
+// LazyDecodeSetting is a key on context that can be used to determine
+// whether decoding should be deferred.
+const LazyDecodeSetting decodeSetting = iota
+
+// ContextWithLazyDecode provides a context that indicates to readers
+// that any decoding should instead return a LazyEncodedImage.
+func ContextWithLazyDecode(ctx context.Context) context.Context {
```
Comment (edaniels) [inline rimage/lazy_encoded.go:40]: delete this?

**Review by edaniels — APPROVED** [review]
lgtm but have some qs

File: components/camera/client_test.go [hunk 8]
```diff
@@ -195,7 +195,8 @@ func TestClient(t *testing.T) {
 		cameraDepthClient, ok := client.(camera.Camera)
 		test.That(t, ok, test.ShouldBeTrue)
 
-		frame, _, err := camera.ReadImage(context.Background(), cameraDepthClient)
+		frame, _, err := camera.ReadImage(
+			gostream.WithMIMETypeHint(context.Background(), rutils.MimeTypePNG), cameraDepthClient)
```
Comment (tessavitabile) [inline components/camera/client_test.go:199]: [nit] This change should no longer be needed.

File: components/camera/server_test.go [hunk 9]
```diff
@@ -166,18 +163,7 @@ func TestServer(t *testing.T) {
 		test.That(t, imageReleased, test.ShouldBeTrue)
 		imageReleasedMu.Unlock()
 		test.That(t, resp.MimeType, test.ShouldEqual, utils.MimeTypeRawRGBA)
-		test.That(t, resp.Image, test.ShouldResemble, img.Pix)
-
-		resp, err = cameraServer.GetImage(
-			context.Background(),
-			&pb.GetImageRequest{Name: testCameraName, MimeType: ""},
```
Comment (tessavitabile) [inline components/camera/server_test.go:173]: We still want to test MimeType:"".

File: components/camera/server_test.go [hunk 10]
```diff
@@ -244,31 +227,11 @@ func TestServer(t *testing.T) {
 		test.That(t, resp.MimeType, test.ShouldEqual, utils.MimeTypePNG)
 		test.That(t, resp.Image, test.ShouldResemble, depthBuf.Bytes())
 		// bad camera
-		_, err = cameraServer.GetImage(context.Background(), &pb.GetImageRequest{Name: failCameraName})
+		_, err = cameraServer.GetImage(context.Background(), &pb.GetImageRequest{Name: failCameraName, MimeType: utils.MimeTypeRawRGBA})
 		test.That(t, err, test.ShouldNotBeNil)
 		test.That(t, err.Error(), test.ShouldContainSubstring, "can't generate stream")
 	})
 
-	t.Run("GetImage with lazy", func(t *testing.T) {
```
Comment (tessavitabile) [inline components/camera/server_test.go:252]: Can this test case be added back?

File: components/camera/server_test.go [hunk 11]
```diff
@@ -298,20 +261,14 @@ func TestServer(t *testing.T) {
 		test.That(t, resp.ContentType, test.ShouldEqual, "image/png")
 		test.That(t, resp.Data, test.ShouldResemble, imgBuf.Bytes())
 
-		imageReleasedMu.Lock()
```
Comment (tessavitabile) [inline components/camera/server_test.go:301]: Is it possible to revert these changes? I don't want to have to figure out how they change the meaning of the tests.

File: components/camera/videosource/helpers.go [hunk 12]
```diff
@@ -15,9 +15,27 @@ import (
...
+		formatMatch := bytes.Compare(data[:4], rimage.RGBABitmapMagicNumber)
+		if formatMatch != 0 {
+			return "", errors.New(
+				"attempted to decode raw rgba data, but data was not encoded with the expected header format")
+		}
+	}
+	if requestedMime == "" {
+		requestedMime = usedMimeType
+	}
+	return requestedMime, nil
```
[the remark this answers is not in the document]
Comment (tessavitabile) [reply components/camera/videosource/helpers.go:39]: I think this is still an issue.

File: components/camera/videosource/helpers.go [hunk 13]
```diff
@@ -44,34 +63,29 @@ func readyBytesFromURL(ctx context.Context, client http.Client, url string) ([]b
...
+			)
+		}
+		return rimage.NewLazyEncodedImage(colorData, detectedMimeType), nil
+	}
+	if actualMime == "" {
+		actualMime = detectedMimeType
 	}
 
-	img, err := decodeImage(colorData)
+	img, err := rimage.DecodeImage(ctx, colorData, actualMime)
```
Comment (tessavitabile) [inline components/camera/videosource/helpers.go:88]: At this point, we still don't know that actualMime==detectedMimeType, so we could be passing the wrong mime type to DecodeImage().

File: components/camera/videosource/helpers.go [hunk 14]
```diff
@@ -83,16 +97,15 @@ func readDepthURL(ctx context.Context, client http.Client, url string, immediate
...
 	}
-
-	if !immediate {
-		if lazyImg, ok, err := checkLazyFromData(ctx, depthData); err != nil {
-			return nil, err
-		} else if ok {
-			return lazyImg, nil
-		}
+	mimeType, err := checkMimeType(ctx, depthData)
+	actualMimeType, isLazy := utils.CheckLazyMIMEType(mimeType)
```
Comment (tessavitabile) [inline components/camera/videosource/helpers.go:101]: err needs to be checked before mimeType is used here

File: components/camera/videosource/server_test.go [hunk 15]
```diff
@@ -90,16 +90,15 @@ func TestServerSource(t *testing.T) {
 	test.That(t, err, test.ShouldBeNil)
 	test.That(t, img, test.ShouldResemble, rimage.ConvertImage(imgDecode))
 
-	lazyCtx := gostream.WithMIMETypeHint(context.Background(), utils.WithLazyMIMEType(utils.MimeTypePNG))
```
Comment (tessavitabile) [inline components/camera/videosource/server_test.go:93]: Can changes in this file be reverted, and instead just remove the extra arguments to NewLazyEncodedImage()?

File: rimage/depth_map.go [hunk 16]
```diff
@@ -117,41 +116,17 @@ func (dm *DepthMap) SubImage(r image.Rectangle) *DepthMap {
 // or if it can be converted into one.
 func ConvertImageToDepthMap(img image.Image) (*DepthMap, error) {
 	switch ii := img.(type) {
-	case *LazyEncodedImage:
```
Comment (tessavitabile) [inline rimage/depth_map.go:120]: Why don't we need to handle the lazy case here?

File: rimage/image_file.go [hunk 17]
```diff
@@ -235,7 +245,7 @@ func EncodeImage(ctx context.Context, img image.Image, mimeType string) ([]byte,
 			return nil, err
 		}
 	default:
-		return nil, errors.Errorf("do not know how to encode %q", actualOutMIME)
+		return nil, errors.Errorf("do not know how to encode %q", mimeType)
```
Comment (tessavitabile) [inline rimage/image_file.go:248]: This should still use actualOutMIME

File: rimage/image_file_test.go [hunk 18]
```diff
@@ -38,15 +38,6 @@ func TestDecodeImage(t *testing.T) {
 
 	var buf bytes.Buffer
 	test.That(t, png.Encode(&buf, img), test.ShouldBeNil)
-
-	t.Run("lazy", func(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (tessavitabile) [reply rimage/image_file_test.go:42]: Can changes in this file be reverted now?

File: rimage/depth_map.go [hunk 19]
```diff
@@ -117,41 +116,17 @@ func (dm *DepthMap) SubImage(r image.Rectangle) *DepthMap {
...
-			return nil, err
-		}
-		return ConvertImageToDepthMap(decoded)
 	case *DepthMap:
 		return ii, nil
 	case *imageWithDepth:
 		return ii.Depth, nil
 	case *image.Gray16:
 		return gray16ToDepthMap(ii), nil
-	case *image.NRGBA64:
```
Comment (tessavitabile) [inline rimage/depth_map.go:133]: Is it okay to get rid of NRGBA64 before reintroducing .dat? CC @bhaney

File: utils/mime.go [hunk 20]
```diff
@@ -25,16 +30,23 @@ const (
...
-
 	// MimeTypeTabular used to indicate tabular data, this is used mainly for filtering data.
 	MimeTypeTabular = "x-application/tabular"
 
 	// MimeTypeDefault used if mimetype cannot be inferred.
 	MimeTypeDefault = "application/octet-stream"
 )
 
+// SupportedMimeTypes is a set of the currently supported MIME types.
+var SupportedMimeTypes rutils.StringSet = rutils.NewStringSet(
```
Comment (tessavitabile) [inline utils/mime.go:41]: This is unused.

File: rimage/depth_map.go (hunk 19 above)
Comment (bhaney) [reply rimage/depth_map.go:133]: Right now, none of our camera servers return a raw depth image encoded in NRGBA64. Depth info is still successfully transmitted as a png (and I think SLAM still is just using png?) 

Given that our new header will only apply for 8-bit color too, I think the 16-bit color "encodings" will become broken if we leave it in. So it's best to remove it.

Comment (tessavitabile) [reply rimage/depth_map.go:133]: Thank you!

File: rimage/image_file.go [hunk 21]
```diff
@@ -22,6 +23,15 @@ import (
 	ut "go.viam.com/rdk/utils"
 )
 
+// RGBABitmapMagicNumber represents the magic number for our custom header
+// for raw RGBA data. See the comments in DecodeImage and EncodeImage for
+// more information.
```
Comment (bhaney) [inline rimage/image_file.go:28]: Can you put the link or description of the format here too? Maybe just write out the description, because I'm afraid of link rot, and the potential of the medium article to go 404 at some point.

**Review by bhaney — COMMENTED** [review]
All of the RGBA format stuff looks good! I think you can remove most of the changes to "lazy" encoding/decoding and keep them as they were, as Tess's comments point out.

**Comment by bhaney** [comment]
Another thing is the Bounds() function in the lazyEncoding struct can be written in such a way to just extract the height/width info from the headers, without having to decode the entire image now

File: components/camera/client.go [hunk 22]
```diff
@@ -60,11 +60,9 @@ func (c *client) Read(ctx context.Context) (image.Image, func(), error) {
 		return nil, nil, err
 	}
 	if actualType != resp.MimeType {
-		c.logger.Debugw("got different MIME type than what was asked for", "sent", actualType, "received", resp.MimeType)
-	} else {
-		resp.MimeType = mimeType
```
[the remark this answers is not in the document]
Comment (gvaradarajan) [reply components/camera/client.go:65]: is the response object used past this point?

**Comment by tessavitabile** [comment]
> Another thing is the Bounds() function in the lazyEncoding struct can be written in such a way to just extract the height/width info from the headers, without having to decode the entire image now

If that's the case, then we don't need to remove height/width from the GetImage response in order to avoid decoding lazy. Should we still make this breaking change?

File: components/camera/client.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (tessavitabile) [reply components/camera/client.go:65]: No, but I realized it should be used on line 65.

File: components/camera/client_test.go (hunk 8 above)
Comment (tessavitabile) [reply components/camera/client_test.go:199]: ah, got it, thanks

File: components/camera/videosource/helpers.go (hunk 13 above)
Comment (gvaradarajan) [reply components/camera/videosource/helpers.go:88]: is this a current concern? right now, we're using image.Decode and that deduces the mime type from the data, not one supplied by us. There are multiple scenarios right now where requestedMime in the existing checkLazyFromData function turns out to be an empty string and the read urls functions don't care and let image.Decode handle it. Should I change the functions to not accept this scenario?

File: components/camera/videosource/helpers.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (gvaradarajan) [reply components/camera/videosource/helpers.go:39]: sorry I missed this last time, but I have some concerns that I outlined in response to the next comment

File: components/camera/videosource/helpers.go (hunk 13 above)
Comment (gvaradarajan) [reply components/camera/videosource/helpers.go:88]: as a side note, I am changing the code and adding comments to at least make that situation clearer for now

File: rimage/depth_map.go (hunk 16 above)
Comment (gvaradarajan) [reply rimage/depth_map.go:120]: we do, when I undid my changes I forgot this part, adding it back

Comment (gvaradarajan) [reply rimage/depth_map.go:120]: adding back the test cases I forgot, so that will make sure I don't forget about this one again

File: components/camera/videosource/helpers.go (hunk 13 above)
Comment (tessavitabile) [reply components/camera/videosource/helpers.go:88]: I didn't quite follow. What I meant was I think we should pass detectedMimeType to DecodeImage here instead of actualMime, since actualMime is the requested mime type and may not be what was returned.

File: components/camera/client.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (gvaradarajan) [reply components/camera/client.go:65]: but won't the response object be garbage collected when the function ends? so why do we need to edit the response object?

[the remark this answers is not in the document]
Comment (gvaradarajan) [reply components/camera/client.go:65]: oh I have finally understood what you meant, I'm sorry it took so long. Will change

File: components/camera/client.go [hunk 23]
```diff
@@ -60,11 +60,9 @@ func (c *client) Read(ctx context.Context) (image.Image, func(), error) {
...
+		c.logger.Debugw("got different MIME type than what was asked for", "sent", mimeType, "received", resp.MimeType)
 	}
-	img, err := rimage.DecodeImage(ctx, resp.Image, resp.MimeType, int(resp.WidthPx), int(resp.HeightPx))
+	img, err := rimage.DecodeImage(ctx, resp.Image, mimeType)
```
[the remark this answers is not in the document]
Comment (gvaradarajan) [reply components/camera/client.go:65]: should Read not be able to return a lazily encoded image?

File: components/camera/client.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (gvaradarajan) [reply components/camera/client.go:65]: well actually I still had a question, but I asked it in reply to the comment on L65

File: components/camera/client.go (hunk 23 above)
[the remark this answers is not in the document]
Comment (tessavitabile) [reply components/camera/client.go:65]: Hmm I'm not sure, since I guess we decode the image here. But my concern is that we are passing in the wrong mime type here (in case GetImage returned a different mime type from the one we requested).

File: components/camera/videosource/helpers.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (gvaradarajan) [reply components/camera/videosource/helpers.go:50]: if I return detectedMimeType here and I use it for the input for image.DecodeImage in readColorURL and readDepthURL, won't it remove the functionality of being able to return a lazy image?

File: components/camera/videosource/helpers.go [hunk 24]
```diff
@@ -15,9 +16,38 @@ import (
...
+			)
+		}
+		formatMatch := bytes.Compare(data[:4], rimage.RGBABitmapMagicNumber)
+		if formatMatch != 0 {
+			return "", errors.New(
+				"attempted to decode raw rgba data, but data was not encoded with the expected header format")
+		}
+	} else if actualMime != "" && actualMime != detectedMimeType {
+		return "", errors.Errorf(
+			"mime type requested (%q) for lazy decode not returned (got %q)",
```
Comment (tessavitabile) [inline components/camera/videosource/helpers.go:38]: We haven't check that a lazy mime type was requested here.

File: components/camera/videosource/helpers.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (tessavitabile) [reply components/camera/videosource/helpers.go:50]: Ah, yes, that's correct. If the requestedMime was lazy, then this should returned the requestedMime (after checking that the detectedMimeType is the non-lazy version of the same mime type). Otherwise, it should return the detectedMimeType. This would handle the case where the requestedMime is png and the detectedMimeType is jpeg, for example. In that case, we don't want to pass png to rimage.DecodeImage(). (I'm assuming rimage.DecodeImage() wants the actual mime type rather than the requested mime type? If not, maybe both this function and that one should have comments stating what mime type represents.)

File: components/camera/videosource/helpers.go [hunk 25]
```diff
@@ -44,34 +74,19 @@ func readyBytesFromURL(ctx context.Context, client http.Client, url string) ([]b
...
+	if err != nil {
 		return nil, err
-	} else if ok {
-		return lazyImg, nil
 	}
-
-	img, err := decodeImage(colorData)
+	img, err := rimage.DecodeImage(ctx, colorData, mimeType)
+	if _, isLazy := utils.CheckLazyMIMEType(mimeType); isLazy {
+		return img, nil
```
Comment (tessavitabile) [inline components/camera/videosource/helpers.go:88]: err must be checked before using img here.

File: components/camera/videosource/helpers.go [hunk 26]
```diff
@@ -83,18 +98,17 @@ func readDepthURL(ctx context.Context, client http.Client, url string, immediate
...
-		}
+	mimeType, err := checkMimeType(ctx, depthData)
+	if err != nil {
+		return nil, err
 	}
-
-	img, err := decodeImage(depthData)
+	actualMimeType, isLazy := utils.CheckLazyMIMEType(mimeType)
+	if !immediate && isLazy {
+		return rimage.NewLazyEncodedImage(depthData, actualMimeType), nil
```
Comment (tessavitabile) [inline components/camera/videosource/helpers.go:107]: [opt] Since DecodeImage will create the lazy encoded image, you could move the check for `!immediate && isLazy` below the call to DecodeImage, and just return img. This would make the function more similar to readColorURL().

File: rimage/image_file.go [hunk 27]
```diff
@@ -22,6 +25,52 @@ import (
...
+	// Here we register the custom format above so that we can simply use image.Decode
+	// so long as the raw RGBA data has the appropriate header
+	image.RegisterFormat("vnd.viam.rgba", string(RGBABitmapMagicNumber),
+		func(r io.Reader) (image.Image, error) {
+			rawBytes, err := ioutil.ReadAll(r)
+			if err != nil {
+				return nil, err
+			}
+			header := rawBytes[:RawRGBAHeaderLength]
+			width := int(binary.BigEndian.Uint32(header[4:8]))
```
Comment (tessavitabile) [inline rimage/image_file.go:49]: I think this can panic if `header` isn't long enough.

File: rimage/image_file.go [hunk 28]
```diff
@@ -160,47 +209,26 @@ func SaveImage(pic image.Image, loc string) error {
...
-		img, err := qoi.Decode(bytes.NewReader(imgBytes))
-		return img, err
-	default:
-		return nil, errors.Errorf("do not how to decode MimeType %s", mimeType)
+	img, usedMimeType, err := image.Decode(bytes.NewReader(imgBytes))
+	if err != nil {
+		return nil, err
 	}
+	if ("image/" + usedMimeType) != mimeType {
+		return nil, errors.Errorf(
```
Comment (tessavitabile) [inline rimage/image_file.go:225]: This is a behavior change from today. Is this going to prevent camera servers from returning a different mime type from the one that was requested?

**Review by tessavitabile — COMMENTED** [review]
Very nearly! Let me know if you'd like to discuss what the contract of checkMimeType() should be.

File: rimage/image_file.go (hunk 28 above)
Comment (gvaradarajan) [reply rimage/image_file.go:225]: uh oh, I didn't intend a behavior change. Um, I think I can just remove?

File: components/camera/videosource/helpers.go (hunk 24 above)
Comment (gvaradarajan) [reply components/camera/videosource/helpers.go:38]: in the event of a lazy mime type being requested, isn't `actualMime` here representing the non-lazy type being requested (e.g. if `requstedMime` is "image/png+lazy", `actualMime` would be "image/png")?

File: rimage/image_file.go (hunk 28 above)
Comment (tessavitabile) [reply rimage/image_file.go:225]: yes, i think so

File: components/camera/videosource/helpers.go (hunk 24 above)
Comment (tessavitabile) [reply components/camera/videosource/helpers.go:38]: I'm saying that we need to check whether a lazy mime type was requested before returning this error .

Comment (gvaradarajan) [reply components/camera/videosource/helpers.go:43]: I put lazy decode in the error statement, which was a mistake. My intention was to do this check regardless of whether the original type requested a LazyEncodedImage. Is that the behavior we want though?

File: rimage/image_file.go [hunk 29]
```diff
@@ -34,6 +36,41 @@ var RGBABitmapMagicNumber = []byte("RGBA")
...
+			img.Pix = imgBytes
+			return img, nil
+		},
+		func(r io.Reader) (image.Config, error) {
+			imgBytes := make([]byte, RawRGBAHeaderLength)
+			_, err := r.Read(imgBytes)
+			if err != nil {
+				return image.Config{}, err
+			}
+			header := imgBytes[:RawRGBAHeaderLength]
```
Comment (bhaney) [inline rimage/image_file.go:62]: Shouldn't the imageBytes already be as long as the Length? Though you may also want to check for a panic here as well if it didn't end up reading enough bytes.

File: rimage/image_file.go [hunk 30]
```diff
@@ -34,6 +36,41 @@ var RGBABitmapMagicNumber = []byte("RGBA")
...
+			width := int(binary.BigEndian.Uint32(header[4:8]))
+			height := int(binary.BigEndian.Uint32(header[8:12]))
+			img := image.NewNRGBA(image.Rect(0, 0, width, height))
+			imgBytes := rawBytes[RawRGBAHeaderLength:]
+			img.Pix = imgBytes
+			return img, nil
+		},
+		func(r io.Reader) (image.Config, error) {
+			imgBytes := make([]byte, RawRGBAHeaderLength)
+			_, err := r.Read(imgBytes)
```
Comment (bhaney) [inline rimage/image_file.go:58]: We don't want to read the bytes of the entire image, just the number up to RawRGBAHeaderLength. Can you check that it only returns the minimum number of bytes necessary?

**Review by bhaney — COMMENTED** [review]
Nice use of RegisterFormat! 

I just have one question - does r.Read() for image.Config only read the specified number of header bytes? We won't want it to read all of the image bytes when it will just use the first set.

File: components/camera/videosource/helpers.go (hunk 24 above)
Comment (gvaradarajan) [reply components/camera/videosource/helpers.go:43]: oh should I change DecodeImage back then? Just calling image.Decode will use the MIME type in the header of the data

File: rimage/image_file.go (hunk 30 above)
Comment (gvaradarajan) [reply rimage/image_file.go:58]: does Read not use the capacity of the byte array to determine how many bytes to read?

Comment (gvaradarajan) [reply rimage/image_file.go:58]: I looked it up, it [does](https://pkg.go.dev/bufio#Reader.Read). However, there may be too few bytes in the original data as @tessavitabile pointed out, so I will check for that and error accordingly

**Review by bhaney — APPROVED** [review]
All the RGBA format stuff is looking good to me!

File: rimage/image_file.go [hunk 31]
```diff
@@ -22,6 +24,55 @@ import (
 	ut "go.viam.com/rdk/utils"
 )
 
+// RGBABitmapMagicNumber represents the magic number for our custom header
+// for raw RGBA data. The header is composed of this magic number followed by
+// a 4-bit line of the width as a uint32 number and another for the height. Credit to
```
Comment (bhaney) [inline rimage/image_file.go:29]: I think you mean 4-byte.  4 bytes = 32 bits

File: components/camera/videosource/helpers.go (hunk 24 above)
Comment (tessavitabile) [reply components/camera/videosource/helpers.go:43]: I think you can leave DecodeImage as is, but pass it the detected mime type.

File: components/camera/videosource/helpers.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (tessavitabile) [reply components/camera/videosource/helpers.go:55]: Can you respond to this before I take another pass? OR do you want to talk about it in person?

File: rimage/image_file.go (hunk 31 above)
Comment (gvaradarajan) [reply rimage/image_file.go:29]: I did

**Review by tessavitabile — APPROVED** [review]
LGTM once Bijan's comment about the panic in the second RegisterFormat function is resolved. Yayyyy!!!!!!!!
